feat(driver): Add support for `encoding` option in cy.request()

hossam-magdy opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/.github/CONTRIBUTING.md 
-->

- Closes #2029 
- Closes #3576 

Originally, the PR https://github.com/cypress-io/cypress/pull/7380 (which describes the issue throughly and includes case analysis) was the attempt to solve the issues in consideration, but it was thought that there might be a different, yet very simple (one-line) solution.

## User facing changelog

Add support for `encoding` option in `cy.request()`

<!--
Explain the change(s) for every user to read in our changelog.
-->

## Additional details

<!--
Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

Introduce/use/support `encoding` option to [`cy.request()`](https://docs.cypress.io/api/commands/request.html), which is already supported by `@cypress/request` package (see [request.js:1130](https://github.com/cypress-io/request/blob/8afbb25/request.js#L1130)).

By introducing this option, the user can specify the `encoding` of the `response.body`. Which means, for instance, that the user can set the `encoding` of request to grab a zip/pdf/audio/video/… file, to `binary`, so that the [`fullBufferOfResponseBody.toString()`](https://github.com/cypress-io/request/blob/8afbb25/request.js#L1130) transformation process is done without resulting into invalid data.

Because the value of this option is eventually passed to `buffer.toString(ENCODING)`, the possible values are the ones mentioned in the [official Node documentation](https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings):
- `utf8`
- `utf16le`
- `latin1`
- `base64`
- `hex`
- `ascii`
- `binary` (an alias of `latin1`)
- `ucs2` (an alias of `utf16le`)

And because the user will have the ability to assign those values. it might be good to consider validating the value before using it in beffer.toString(). For that, a validation and sanitization (lowercasing) was implemented. Also as extra validation (which might be thought as unnecessary, up to reviewers!), a PR in @cypress/request https://github.com/cypress-io/request/pull/5 was created as well.

This solution was tested with the code shown below that the exact file requested with the exact size is downloaded and written to the filesystem exactly as expected.

## How has the user experience changed?

<!--
Provide before and after examples of the change.
Screenshots or GIFs are preferred.
-->

By setting the `encoding` option in `cy.request()` to binary, the downloaded files from the following sample test are valid, and exactly the same as expected. Like [this](https://github.com/cypress-io/cypress/commit/55aff0074e):

```js
describe('Download files by `cy.request()`->`cy.writeFile()`', () => {
  const urls = [
    'https://file-examples.com/wp-content/uploads/2017/02/zip_2MB.zip',
    'https://file-examples.com/wp-content/uploads/2017/10/file_example_PNG_500kB.png',
    'https://file-examples.com/wp-content/uploads/2017/02/file-sample_100kB.doc',
    'https://file-examples.com/wp-content/uploads/2017/02/file-sample_100kB.docx',
    'https://file-examples.com/wp-content/uploads/2017/02/file_example_XLS_50.xls',
    'https://file-examples.com/wp-content/uploads/2017/02/file_example_XLSX_1000.xlsx',
    'https://file-examples.com/wp-content/uploads/2017/10/file-example_PDF_500_kB.pdf',
  ];

  for (const url of urls) {
    const ext = url.substr(url.lastIndexOf('.') + 1);

    it(`Download ${ext.toUpperCase()} file`, () => {
      cy.request({
        url,
        encoding: 'binary', // !IMPORTANT!
      }).then((response) => {
        const filePath =
          Cypress.config('screenshotsFolder') + `/${ext}-downloaded.${ext}`;
        const expectedFileSize = response.headers['content-length'];
        console.log({ ext, size: expectedFileSize, filePath });
        // !IMPORTANT!: notice `encoding: 'binary'` below
        cy.writeFile(filePath, response.body, { encoding: 'binary' });
        cy.exec(`ls -al ${filePath} | awk '{print $5}'`).then(
          ({ stdout: actualFileSize }) => {
            expect(actualFileSize).to.equal(expectedFileSize);
          }
        );
      });
    });
  }
});
```

<img width="449" alt="Screenshot 2020-05-17 at 22 56 19" src="https://user-images.githubusercontent.com/17128077/82159880-d0dc4d80-9891-11ea-90a8-5b39116c1876.png">

Also a simple example like [this one](https://github.com/cypress-io/cypress/issues/2029#issuecomment-400508816), will be working if `encoding` is set to `binary` in _both_ `cy.request` and `cy.writeFile`

---

## Analysis of the problem's root cause

`cy.request()` is using `utf8` as the encoding value of `responseBuffer->string` transformation. How?! In [the list of options](https://github.com/cypress-io/cypress/blob/a27bd34/packages/driver/src/cy/commands/request.js#L17-L32) passed to `@cypress/request`, there is no `encoding`. Although `@cypress/request` actually [supports the encoding option](https://github.com/cypress-io/request/blob/8afbb25/request.js#L1129-L1131) during the readResponseBody process, which, [if the response body is a buffer](https://github.com/cypress-io/request/blob/8afbb25/request.js#L1108-L1112), boils down to a simple buffer->string transformation.

This analysis is about **using different encoding values (utf8, binary, …) in a `buffer->string->buffer` 2-way transformation of a sample PDF file (with size of _66851 bytes_)**, which is relevant to issue: https://github.com/cypress-io/cypress/issues/3576.

Explanation of the analysis stages:

**Stage 1**: `buffer->string`: in the heart of `cy.request()`, during the transformation of the response.body buffer into string. Where the logged values are:

- full buffer size (as received from request)
- first byte in the buffer
- last byte in the buffer
- size of string resulted from the buffer->string transformation (through `buffer.toString(ENCODING)`)

**Stage 2**: `string->buffer`: in the heart of `cy.writeFile()`, during the transformation of the string of file content passed to `cy.writeFile()` into buffer. The logged values in this stage are:

- size of the string (the content value passed to `cy.writeFile()`)
- size of buffer resulted from the string->buffer transformation (through `Buffer.from(content, ENCODING)`)
- first byte in the buffer
- last byte in the buffer

| Criterion                                                                                          | binary<br>[latin1] | ascii | base64 | hex    | utf16le<br>[ucs2] | utf8       |
| -------------------------------------------------------------------------------------------------- | ------------------ | ----- | ------ | ------ | ----------------- | ---------- |
| Full buffer size in stage 1                                                                        | 66851              | 66851 | 66851  | 66851  | 66851             | 66851      |
| Full buffer size in stage 2                                                                        | 66851              | 66851 | 66851  | 66851  | **66850**         | **101377** |
| First byte in the buffer of stage 1                                                                | 37                 | 37    | 37     | 37     | 37                | 37         |
| First byte in the buffer of stage 2                                                                | 37                 | 37    | 37     | 37     | 37                | 37         |
| First byte in the buffer of stage 1                                                                | 70                 | 70    | 70     | 70     | 70                | 70         |
| First byte in the buffer of stage 2                                                                | 70                 | 70    | 70     | 70     | 70                | 70         |
| Size of string value in stage 1<br>(after responseBuffer->string transformation)                   | 66851              | 66851 | 89136  | 133702 | 33425             | 64703      |
| Size of string value in stage 2<br>(before content->buffer transformation)                         | 66851              | 66851 | 89136  | 133702 | 33425             | 64703      |
| Percentage of string size / buffer size<br>(more data transfer between cypress client and server?) | 100%               | 100%  | 133.3% | 100%   | 50%               | 96.7%      |
| PDF file seems to have the expected content (PDF viewer / Chrome)                                  | **✓**              | **✓** | **✓**  | **✓**  | **✓**             | **✗**      |
| Both stages has the exact same buffer value. I.e: valid PDF result                                                        | **✓**              | **✓** | **✓**  | **✓**  | **✗**             | **✗**      |

### Analysis conclusion

In Node.Js, a 2-way transformation (buffer->string->buffer) of a PDF file content (i.e: binary) is likely to work as expected using encoding: `binary` (alias: `latin1`), `ascii`, `base64`, or `hex`. 

Using `utf16le` (alias: `ucs2`) or `utf8` leads to **invalid** content. Which might also be reasonably expected by reading [NodeJs documentation](https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings) in case of invalid utf8 data (i.e: binary data).

The guaranteed results are from `binary`, as expected for a binary file!. And the mid-way string size is exactly the same as the buffer size before and after the transformation.

And because in cypress (by the time of this writing and as of [v4.6.0](https://github.com/cypress-io/cypress/releases/tag/v4.6.0)): `cy.request()` is executed through the cypress driver in NodeJs environment, and defaults to `utf8` encoding option under the hood, the resulted PDF content is invalid.

---

## PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [x] Have tests been added/updated?
- [ ] (N/A) Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [x] Has a PR for user-facing changes been opened in https://github.com/cypress-io/cypress-documentation/pull/2805
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [ ] (N/A) Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

hossam-magdy commented:
@jennifer-shehane thanks for your comment. I still need to wrap my head around how things are working in the codebase, especially in the tests part. I will try to invest sometime soon for that!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=7382) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/cypress-io/cypress?pullRequest=7382) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: hossam-magdy<br/>:x: Hossam Magdy<br/><hr/>**Hossam Magdy** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cypress-io/cypress?pullRequest=7382) it.</sub>

hossam-magdy commented:
@jennifer-shehane The PR is ready. I added the tests, and rebased to the latest `develop` changes. Also updated https://github.com/cypress-io/cypress-documentation/pull/2805.

hossam-magdy commented:
@flotwig Thanks!

That part you quoted from the PR description was written before implementing the validation in this PR (I should have removed it from the description, doing it now). And I also thought that crashing of the whole process is not convenient, so I created the other PR.

But then I implemented the validation as well as the sanitization in this PR. So I agree with you that the other PR (the one in the request package) is not necessary.

Hrebeniuk1 commented:
I have cypress version 9.6.1 and this is not gonna work, also with code example from this pr and 2 links from the beginning. Is there a workaround of this problem? How to cy.writeFile() pdf file?
